### PR TITLE
Prototype build from source

### PR DESCRIPTION
**Pull request by gearama**
Adds new cmake file to define macros that download deps , adds them to the list of folders that will be built, defines the build configs. 

all changes are behind build option. there are a couple of dockerfiles as examples. 

I ended up going with source code directly as vcpkg seems to not fit well to this usecase
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: cmake-modules/FolderList.cmake [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    # need a nicer way to copy/move folder 
+    # this is stupid, i need to archive the folder then extract at new location
+    execute_process(COMMAND tar -cf  ${DOWNLOAD_FOLDER}/${DEP_NAME}.tar -C ${DOWNLOAD_FOLDER}/${DEP_NAME}/${DEP_PREFIX}-${DEP_NAME}_${DEP_VERSION}/${DEP_FOLDER} .)
+    file(ARCHIVE_EXTRACT INPUT ${DOWNLOAD_FOLDER}/${DEP_NAME}.tar DESTINATION ${CMAKE_SOURCE_DIR}/${DEP_FOLDER})
+    #cleanup
+    file(REMOVE_RECURSE ${DOWNLOAD_FOLDER})
+    #add dependency folder to build list
+    list(APPEND BUILD_FOLDERS ${DEP_FOLDER})
+
+endmacro()
```
Comment (vhvb1989) [inline cmake-modules/FolderList.cmake:57]: neew line?

File: cmake-modules/FolderList.cmake [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    set(DEP_PREFIX azure-sdk-for-cpp)
+    # get the zip
+    file(DOWNLOAD https://github.com/Azure/azure-sdk-for-cpp/archive/refs/tags/${DOWNLOAD_FILE} ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE})
+
+    #extract the zip
+    file(ARCHIVE_EXTRACT INPUT ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE} DESTINATION ${DOWNLOAD_FOLDER}/${DEP_NAME})
+    #make target folder
+    file(MAKE_DIRECTORY ${CMAKE_SOURCE_DIR}/${DEP_FOLDER})
+    # need a nicer way to copy/move folder 
+    # this is stupid, i need to archive the folder then extract at new location
```
Comment (vhvb1989) [inline cmake-modules/FolderList.cmake:49]: we might not want to keep the `stupid` there :)

File: cmake-modules/FolderList.cmake [hunk 3]
```diff
@@ -0,0 +1,57 @@
+macro(GetFolderList project)
+    message ("project found ${project}")
+  
+    if(${project} STREQUAL CERTIFICATES)
+        DownloadDepVersion(sdk/core azure-core 1.2.1)
+        DownloadDepVersion(sdk/identity azure-identity 1.1.0)
+        #list(APPEND BUILD_FOLDERS "sdk/keyvault/azure-security-keyvault-certificates")
```
Comment (vhvb1989) [inline cmake-modules/FolderList.cmake:7]: remove?

File: sdk/keyvault/azure-security-keyvault-certificates/CMakeLists.txt [hunk 4]
```diff
@@ -1,103 +1,124 @@
...
+include(AzureConfigRTTI)
+
+az_vcpkg_integrate()
+
+if(BUILD_SOURCE_DEPS)
+    GetFolderList(CERTIFICATES)
+    foreach(oneFolder IN LISTS BUILD_FOLDERS)
+	    message("add folder ${oneFolder}")
+	    add_subdirectory(${oneFolder})
+        include_directories(${oneFolder})
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-certificates/CMakeLists.txt:37]: Isn't `add_subdirectory` enough?  Why do we want to include the src dir as part of the library we are building?

File: sdk/keyvault/azure-security-keyvault-certificates/CMakeLists.txt [hunk 5]
```diff
@@ -1,103 +1,124 @@
...
+)
+add_library(Azure::azure-security-keyvault-certificates ALIAS azure-security-keyvault-certificates)
+
+target_include_directories(
+  azure-security-keyvault-certificates
+    PUBLIC
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/inc>
+      $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/../../../sdk/core/inc>
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/sdk/core/azure-core/inc>
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/sdk/core/azure-core-test/inc>
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-certificates/CMakeLists.txt:74]: why?  cmake should build core and any other linked target before building this library. This is not cool

File: sdk/keyvault/azure-security-keyvault-certificates/CMakeLists.txt [hunk 6]
```diff
@@ -1,103 +1,124 @@
...
+message($<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/inc)
+# Include shared source code
+# NOTE: Use shared-code only within .cpp files. DO NEVER consume the shared-code from header files.
+target_include_directories(
+  azure-security-keyvault-certificates
+    PRIVATE
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../azure-security-keyvault-shared/inc>
+)
+
+target_link_libraries(azure-security-keyvault-certificates PUBLIC Azure::azure-core)
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-certificates/CMakeLists.txt:86]: This should make CMake to find Azure::azure-core as a target source and built it first

File: sdk/keyvault/azure-security-keyvault-certificates/CMakeLists.txt [hunk 7]
```diff
@@ -1,103 +1,124 @@
...
+    "azure/keyvault/certificates/dll_import_export.hpp"
+  )
+
+az_rtti_setup(
+  azure-security-keyvault-certificates
+  SECURITY_KEYVAULT_CERTIFICATES
+  "azure/keyvault/certificates/rtti.hpp"
+)
+
+unset(BUILD_SOURCE_DEPS CACHE)
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-certificates/CMakeLists.txt:124]: new line

File: sdk/keyvault/azure-security-keyvault-certificates/CMakeSettings.json [hunk 8]
```diff
@@ -0,0 +1,210 @@
+﻿{
+  "configurations": [
+    {
+      "name": "x64-DebugWithTests",
+      "generator": "Ninja",
+      "configurationType": "Debug",
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-certificates/CMakeSettings.json:6]: is this VS exclusive?  Should we keep it ignored?

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,85 @@
+FROM debian:10
```
Comment (vhvb1989) [inline Dockerfile:1]: where and what are you using this for?  Don't we have it already inside https://github.com/Azure/azure-sdk-for-cpp/tree/main/.devcontainer ??

**Review by vhvb1989 — COMMENTED** [review]
I was talking a few days ago about this with @antkmsft 

There's a problem from getting the source code from `/archive/refs/tags`  

If I clone latest `main`, it might happen that the library that I want to build has a dependency on the latest `core` (which is not yet released, so there's not tag for it).  Then, the library will fail even after getting the last released tag.   Should we always get latest `main` instead?   ( Assuming that we will never have breaking changes)

File: sdk/keyvault/azure-security-keyvault-certificates/CMakeSettings.json (hunk 8 above)
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-certificates/CMakeSettings.json:6]: it's in the root folder and not ignored there that's why i added it here

File: Dockerfile (hunk 9 above)
Comment (gearama) [reply Dockerfile:1]: the thing here is that in order to build a docker image it has to be in the root folder of the "build context" aka the root folder where you are targeting. There are some small changes from the dev docker -- no vcpkg -- 
This image can then be used as base for the Docker.src image. I split it in two files in order to speed up testing build docker image

File: .gitignore [hunk 10]
```diff
@@ -340,3 +340,6 @@ build/
 # vscode
 .vscode/
 .factorypath
+
+sdk/**/sdk/
+sdk/**/out/
```
Comment (vhvb1989) [inline .gitignore:345]: You might want to update the `# Build results` in the top of this file instead.  Make `out` to be something like `**/out/**` 
We can also add a comment about why we need it.

File: Dockerfile (hunk 9 above)
Comment (vhvb1989) [reply Dockerfile:1]: >  it has to be in the root folder

I am not super sure about this.

But, I would like to understand better what we want to do with this file and docker.  Can you elaborate about it?  Or maybe create a doc for users/customers for how they can use this?

I am a bit confused if this is only used to test what you are enabling, or if this is actually required/part of the feature.

File: sdk/keyvault/azure-security-keyvault-certificates/CMakeSettings.json (hunk 8 above)
Comment (vhvb1989) [reply sdk/keyvault/azure-security-keyvault-certificates/CMakeSettings.json:6]: Is there a way to use the one from the root folder everywhere?

Otherwise, we would have to maintain multiple files like this.. ( one per package )

File: sdk/keyvault/azure-security-keyvault-certificates/src/certificate_client.cpp [hunk 11]
```diff
@@ -1,497 +1,497 @@
...
-  return Azure::Response<KeyVaultCertificate>(std::move(value), std::move(rawResponse));
-}
-
-const ServiceVersion ServiceVersion::V7_2("7.2");
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/keyvault/certificates/certificate_client.hpp"
+
+#include "azure/keyvault/shared/keyvault_shared.hpp"
```
Comment (vhvb1989) [inline sdk/keyvault/azure-security-keyvault-certificates/src/certificate_client.cpp:6]: The changes to this file is only a header re-order. It's not relevant to this PR. Can you revert the change to this file?

**Review by vhvb1989 — APPROVED** [review]
Add a last comment to revert changes to certificate_client.cpp.  Please do before merging.

Also, do we have issues to apply this approach to all other SDK lib packages?

File: cmake-modules/FolderList.cmake [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        DownloadDepVersion(sdk/identity azure-identity 1.1.0)
+    endif()
+
+    list(REMOVE_DUPLICATES BUILD_FOLDERS)
+endmacro()
+
+macro(SetCompileOptions project)
+    message ("setting up compile options for ${project}")
+    
+    if(${project} STREQUAL CERTIFICATES)
```
Comment (vhvb1989) [inline cmake-modules/FolderList.cmake:15]: Do you still need this condition if you are making this a general MACRO?

File: sdk/keyvault/ci.yml [hunk 13]
```diff
@@ -44,6 +44,10 @@ stages:
         - Name: azure-security-keyvault-certificates
           Path: azure-security-keyvault-certificates
           VcpkgPortName: azure-security-keyvault-certificates-cpp
+      ArtifactsSource:
+        - Name: azure-security-keyvault-certificates
+          Path: azure-security-keyvault-certificates
+          VcpkgPortName: azure-security-keyvault-keys-cpp
```
Comment (vhvb1989) [inline sdk/keyvault/ci.yml:50]: Remove this. It's duplicated.  Let's use the same artifact

File: sdk/keyvault/ci.yml [hunk 14]
```diff
@@ -65,3 +69,6 @@ stages:
           Value: '-DBUILD_TESTING=ON -DBUILD_SAMPLES=ON'
         - Name: Performance
           Value: '-DBUILD_TESTING=ON -DBUILD_SAMPLES=ON -DBUILD_PERFORMANCE_TESTS=ON'
+      CMakeSourceTestOptions:
```
Comment (vhvb1989) [inline sdk/keyvault/ci.yml:72]: Remove this.
Use `CMakeTestOptions` to add the option

File: eng/pipelines/templates/jobs/archetype-sdk-cmake-generate.yml [hunk 15]
```diff
@@ -72,3 +78,64 @@ jobs:
           CmakeGeneratePath: sdk/${{ parameters.ServiceDirectory }}/${{ artifact.Path }}
           GenerateArgs: ${{ cmakeOption.Value }}
           Env: "$(CmakeEnvArg)"
+
+- job: CMakeSourceGenerate
```
Comment (vhvb1989) [inline eng/pipelines/templates/jobs/archetype-sdk-cmake-generate.yml:82]: We don't need to duplicate this.  You should not need to update this file at all.  Just adding the option to the ci.yml should be enough

File: cmake-modules/FolderList.cmake (hunk 12 above)
Comment (gearama) [reply cmake-modules/FolderList.cmake:15]: i'm leaving it in for now, one i expand it to the other services some options might vary, should not but you never know

File: sdk/keyvault/ci.yml (hunk 14 above)
Comment (gearama) [reply sdk/keyvault/ci.yml:72]: the thing here is that i need a new set of things because the default images for the pipeline are contaminated by vcpkg's dark presence, i need to have a pure, untainted set of machines.

Comment (vhvb1989) [reply sdk/keyvault/ci.yml:72]: > the pipeline are contaminated by vcpkg's dark presence

We want to know that even when there's VCPKG with dependencies, if I set the CMake option, we will get the source instead of using VCPKG.

The CMake option (when == ON ) should prevent CMake from looking the dependency from VCPKG and instead consume it from the source we fetch and build.
